File: internal/analyzers/common/formatter.go
package common

import (
	"fmt"
	"sort"
	"strings"

	"github.com/jedib0t/go-pretty/v6/table"

	"github.com/Sumatoshi-tech/codefang/internal/analyzers/analyze"
)

const (
	percentageValue      = 100
	scoreThresholdHigh   = 0.8
	scoreThresholdMedium = 0.6
)

const msgNoReportData = "No report data available"

// FormatConfig defines configuration for formatting.
type FormatConfig struct {
	SortBy           string
	SortOrder        string
	MaxItems         int
	ShowProgressBars bool
	ShowTables       bool
	ShowDetails      bool
	SkipHeader       bool
}

// Formatter provides generic formatting capabilities for analysis results.
type Formatter struct {
	config FormatConfig
}

// NewFormatter creates a new Formatter with configurable formatting settings.
func NewFormatter(config FormatConfig) *Formatter {
	return &Formatter{
		config: config,
	}
}

// FormatReport formats an analysis report for display.
func (f *Formatter) FormatReport(report analyze.Report) string {
	if report == nil {
		return msgNoReportData
	}

	var parts []string

	// Add header (unless skipped).
	if !f.config.SkipHeader {
		if analyzerName, ok := report["analyzer_name"].(string); ok {
			parts = append(parts, fmt.Sprintf("=== %s ===", strings.ToUpper(analyzerName)))
		}
	}

	// Add summary.
	summary := f.formatSummary(report)
	if summary != "" {
		parts = append(parts, summary)
	}

	// Add progress bars if enabled.
	if f.config.ShowProgressBars {
		progressBars := f.formatProgressBars(report)
		if progressBars != "" {
			parts = append(parts, progressBars)
		}
	}

	// Add tables if enabled.
	if f.config.ShowTables {
		tables := f.formatTables(report)
		if tables != "" {
			parts = append(parts, tables)
		}
	}

	// Add details if enabled.
	if f.config.ShowDetails {
		details := f.formatDetails(report)
		if details != "" {
			parts = append(parts, details)
		}
	}

	return strings.Join(parts, "\n\n")
}

// formatSummary formats the summary section of a report.
func (f *Formatter) formatSummary(report analyze.Report) string {
	var summary []string

	// Add message.
	if message, ok := report["message"].(string); ok && message != "" {
		summary = append(summary, message)
	}

	// Add key metrics.
	metrics := f.extractMetrics(report)
	if len(metrics) > 0 {
		metricLines := make([]string, 0, len(metrics))
		for key, value := range metrics {
			metricLines = append(metricLines, fmt.Sprintf("%s: %.2f", key, value))
		}

		sort.Strings(metricLines)
		summary = append(summary, strings.Join(metricLines, " | "))
	}

	return strings.Join(summary, "\n")
}

// formatProgressBars formats progress bars for numeric values.
func (f *Formatter) formatProgressBars(report analyze.Report) string {
	var bars []string

	// Define count metrics that should not be shown as progress bars.
	countMetrics := map[string]bool{
		"total_comments":        true,
		"good_comments":         true,
		"bad_comments":          true,
		"total_functions":       true,
		"documented_functions":  true,
		"total_comment_details": true,
	}

	// Find numeric values that could be scores (0-1 range).
	for key, value := range report {
		// Skip count metrics.
		if countMetrics[key] {
			continue
		}

		if score, ok := ToFloat64(value); ok && score >= 0 && score <= 1 {
			bar := f.createProgressBar(key, score)
			if bar != "" {
				bars = append(bars, bar)
			}
		}
	}

	if len(bars) == 0 {
		return ""
	}

	return "Progress:\n" + strings.Join(bars, "\n")
}

// formatTables formats data as tables using go-pretty.
func (f *Formatter) formatTables(report analyze.Report) string {
	var tables []string

	// Format collections as tables.
	for key, value := range report {
		if collection, ok := value.([]map[string]any); ok && len(collection) > 0 {
			tableStr := f.formatCollectionTable(key, collection)
			if tableStr != "" {
				tables = append(tables, tableStr)
			}
		}
	}

	return strings.Join(tables, "\n\n")
}

// formatDetails formats detailed information.
func (f *Formatter) formatDetails(report analyze.Report) string {
	var details []string

	// Add all non-collection fields.
	for key, value := range report {
		if _, ok := value.([]map[string]any); !ok {
			details = append(details, fmt.Sprintf("%s: %v", key, value))
		}
	}

	if len(details) == 0 {
		return ""
	}

	sort.Strings(details)

	return "Details:\n" + strings.Join(details, "\n")
}

// formatCollectionTable formats a collection as a table using go-pretty.
func (f *Formatter) formatCollectionTable(collectionKey string, collection []map[string]any) string {
	if len(collection) == 0 {
		return ""
	}

	// Limit items if configured.
	if f.config.MaxItems > 0 && len(collection) > f.config.MaxItems {
		collection = collection[:f.config.MaxItems]
	}

	// Sort if configured.
	if f.config.SortBy != "" {
		f.sortCollection(collection, f.config.SortBy, f.config.SortOrder)
	}

	// Get all unique keys from all items.
	keys := f.getCollectionKeys(collection)
	if len(keys) == 0 {
		return ""
	}

	// Create go-pretty table.
	tbl := table.NewWriter()
	tbl.SetStyle(table.StyleLight)
	tbl.Style().Options.SeparateRows = false
	tbl.Style().Options.SeparateColumns = false
	tbl.Style().Options.DrawBorder = false
	tbl.Style().Options.SeparateHeader = false

	// Add header.
	header := make([]any, len(keys))
	for i, key := range keys {
		header[i] = key
	}

	tbl.AppendHeader(header)

	// Add rows.
	for _, item := range collection {
		row := make([]any, len(keys))
		for i, key := range keys {
			value := item[key]
			if value == nil {
				row[i] = ""
			} else {
				row[i] = fmt.Sprintf("%v", value)
			}
		}

		tbl.AppendRow(row)
	}

	// Add footer with count.
	tbl.AppendFooter(table.Row{fmt.Sprintf("Total: %d items", len(collection))})

	return fmt.Sprintf("%s:\n%s", collectionKey, tbl.Render())
}

// createProgressBar creates a progress bar for a score.
func (f *Formatter) createProgressBar(label string, score float64) string {
	const barLength = 20

	filled := int(score * barLength)
	empty := barLength - filled

	bar := strings.Repeat("█", filled) + strings.Repeat("░", empty)
	percentage := score * percentageValue

	status := "🔴 Poor"
	if score >= scoreThresholdHigh {
		status = "🟢 Good"
	} else if score >= scoreThresholdMedium {
		status = "🟡 Fair"
	}

	return fmt.Sprintf("%s: [%s] %.1f%% %s", label, bar, percentage, status)
}

// extractMetrics extracts numeric metrics from a report.
func (f *Formatter) extractMetrics(report analyze.Report) map[string]float64 {
	metrics := make(map[string]float64)

	for key, value := range report {
		if score, ok := ToFloat64(value); ok {
			metrics[key] = score
		}
	}

	return metrics
}

// getCollectionKeys gets all unique keys from a collection.
func (f *Formatter) getCollectionKeys(collection []map[string]any) []string {
	keySet := make(map[string]bool)

	for _, item := range collection {
		for key := range item {
			keySet[key] = true
		}
	}

	keys := make([]string, 0, len(keySet))
	for key := range keySet {
		keys = append(keys, key)
	}

	sort.Strings(keys)

	return keys
}

// sortCollection sorts a collection by a specific key.
func (f *Formatter) sortCollection(collection []map[string]any, sortBy, sortOrder string) {
	sort.Slice(collection, func(i, j int) bool {
		valI := collection[i][sortBy]
		valJ := collection[j][sortBy]

		// Convert to comparable values.
		compI := f.toComparable(valI)
		compJ := f.toComparable(valJ)

		if sortOrder == "desc" {
			return compI > compJ
		}

		return compI < compJ
	})
}

// toComparable converts a value to a comparable type for sorting.
func (f *Formatter) toComparable(value any) float64 {
	switch typedVal := value.(type) {
	case float64:
		return typedVal
	case int:
		return float64(typedVal)
	case int32:
		return float64(typedVal)
	case int64:
		return float64(typedVal)
	case string:
		return float64(len(typedVal)) // Sort strings by length as fallback.
	default:
		return 0
	}
}

File: rust/crates/cf-analyzers-common/src/formatter.rs
//! Human-readable report formatting.
//!
//! Port of `formatter.go`. [`Formatter`] renders an analyzer [`Report`] as a
//! text block: an optional header, a summary of message + numeric metrics,
//! optional unicode progress bars, optional tables, and optional sorted details.
//!
//! # Byte-identity
//!
//! Per DESIGN §2.7 the `go-pretty` `StyleLight` table output is explicitly
//! **NON-BINDING / cosmetic** — byte-identity is *not* required. The table
//! writer here ([`render_style_light_table`]) reproduces the structure and the
//! byte-width padding behaviour of the Go `StyleLight` style (column dividers,
//! header underline, footer) on a best-effort basis. The machine JSON path lives
//! in [`crate::reporter`] and routes through `cf-gojson`, not through this module.

use crate::report::{Item, Report, Value};
use std::collections::BTreeMap;

/// Message returned for a nil report, mirroring the Go `msgNoReportData`.
pub(crate) const MSG_NO_REPORT_DATA: &str = "No report data available";

const PERCENTAGE_VALUE: f64 = 100.0;
const SCORE_THRESHOLD_HIGH: f64 = 0.8;
const SCORE_THRESHOLD_MEDIUM: f64 = 0.6;

/// Configuration for a [`Formatter`].
///
/// Field names and defaults mirror the Go `FormatConfig` struct.
#[derive(Debug, Clone, Default)]
pub struct FormatConfig {
    /// Key to sort collection rows by (empty = no sorting).
    pub sort_by: String,
    /// Sort direction: `"desc"` for descending, anything else ascending.
    pub sort_order: String,
    /// Maximum number of rows to render per collection (0 = unlimited).
    pub max_items: usize,
    /// Render unicode progress bars for 0–1 numeric scores.
    pub show_progress_bars: bool,
    /// Render collections as tables.
    pub show_tables: bool,
    /// Render a sorted details section of all non-collection fields.
    pub show_details: bool,
    /// Suppress the `=== NAME ===` header.
    pub skip_header: bool,
}

/// Formats analyzer reports for human-readable (text) display.
///
/// Mirrors the Go `Formatter`.
#[derive(Debug, Clone, Default)]
pub struct Formatter {
    config: FormatConfig,
}

impl Formatter {
    /// Creates a formatter with the given configuration. Mirrors `NewFormatter`.
    pub fn new(config: FormatConfig) -> Self {
        Formatter { config }
    }

    /// Formats a report for display. Mirrors `FormatReport`.
    ///
    /// A `None` report yields [`MSG_NO_REPORT_DATA`]. Sections are joined by a
    /// blank line (`"\n\n"`), exactly like the Go implementation.
    pub fn format_report(&self, report: Option<&Report>) -> String {
        let report = match report {
            Some(r) => r,
            None => return MSG_NO_REPORT_DATA.to_string(),
        };

        let mut parts: Vec<String> = Vec::new();

        // Header (unless skipped).
        if !self.config.skip_header {
            if let Some(Value::String(name)) = report.get("analyzer_name") {
                parts.push(format!("=== {} ===", name.to_uppercase()));
            }
        }

        let summary = self.format_summary(report);
        if !summary.is_empty() {
            parts.push(summary);
        }

        if self.config.show_progress_bars {
            let bars = self.format_progress_bars(report);
            if !bars.is_empty() {
                parts.push(bars);
            }
        }

        if self.config.show_tables {
            let tables = self.format_tables(report);
            if !tables.is_empty() {
                parts.push(tables);
            }
        }

        if self.config.show_details {
            let details = self.format_details(report);
            if !details.is_empty() {
                parts.push(details);
            }
        }

        parts.join("\n\n")
    }

    /// Formats the message + sorted numeric metrics summary.
    fn format_summary(&self, report: &Report) -> String {
        let mut summary: Vec<String> = Vec::new();

        if let Some(Value::String(message)) = report.get("message") {
            if !message.is_empty() {
                summary.push(message.clone());
            }
        }

        let metrics = extract_all_numeric_metrics(report);
        if !metrics.is_empty() {
            // BTreeMap iterates sorted by key, then we format and sort the
            // resulting "key: value" lines to match Go's sort.Strings.
            let mut metric_lines: Vec<String> = metrics
                .iter()
                .map(|(key, value)| format!("{key}: {value:.2}"))
                .collect();
            metric_lines.sort();
            summary.push(metric_lines.join(" | "));
        }

        summary.join("\n")
    }

    /// Formats progress bars for numeric values in the 0–1 range.
    fn format_progress_bars(&self, report: &Report) -> String {
        // Count metrics that must not be shown as progress bars.
        const COUNT_METRICS: &[&str] = &[
            "total_comments",
            "good_comments",
            "bad_comments",
            "total_functions",
            "documented_functions",
            "total_comment_details",
        ];

        let mut bars: Vec<String> = Vec::new();
        for (key, value) in report {
            if COUNT_METRICS.contains(&key.as_str()) {
                continue;
            }
            if let Some(score) = value.to_float64() {
                if (0.0..=1.0).contains(&score) {
                    let bar = self.create_progress_bar(key, score);
                    if !bar.is_empty() {
                        bars.push(bar);
                    }
                }
            }
        }

        if bars.is_empty() {
            return String::new();
        }

        format!("Progress:\n{}", bars.join("\n"))
    }

    /// Formats every collection in the report as a table.
    fn format_tables(&self, report: &Report) -> String {
        let mut tables: Vec<String> = Vec::new();
        for (key, value) in report {
            if let Value::Collection(collection) = value {
                if !collection.is_empty() {
                    let table_str = self.format_collection_table(key, collection);
                    if !table_str.is_empty() {
                        tables.push(table_str);
                    }
                }
            }
        }
        tables.join("\n\n")
    }

    /// Formats all non-collection fields as a sorted details section.
    fn format_details(&self, report: &Report) -> String {
        let mut details: Vec<String> = Vec::new();
        for (key, value) in report {
            if !matches!(value, Value::Collection(_)) {
                details.push(format!("{key}: {value}"));
            }
        }

        if details.is_empty() {
            return String::new();
        }

        details.sort();
        format!("Details:\n{}", details.join("\n"))
    }

    /// Formats a single collection as a `StyleLight` table.
    fn format_collection_table(&self, collection_key: &str, collection: &[Item]) -> String {
        if collection.is_empty() {
            return String::new();
        }

        // Limit rows if configured.
        let mut rows: Vec<&Item> = collection.iter().collect();
        if self.config.max_items > 0 && rows.len() > self.config.max_items {
            rows.truncate(self.config.max_items);
        }

        // Sort if configured.
        if !self.config.sort_by.is_empty() {
            self.sort_collection(&mut rows);
        }

        let keys = collection_keys(&rows);
        if keys.is_empty() {
            return String::new();
        }

        let header: Vec<String> = keys.clone();

        let body: Vec<Vec<String>> = rows
            .iter()
            .map(|item| {
                keys.iter()
                    .map(|key| match item.get(key) {
                        None | Some(Value::Null) => String::new(),
                        Some(v) => v.to_string(),
                    })
                    .collect()
            })
            .collect();

        let footer = format!("Total: {} items", rows.len());

        let table = render_style_light_table(&header, &body, &footer);
        format!("{collection_key}:\n{table}")
    }

    /// Creates a unicode progress bar for a score in 0–1.
    fn create_progress_bar(&self, label: &str, score: f64) -> String {
        const BAR_LENGTH: i64 = 20;

        let filled = (score * BAR_LENGTH as f64) as i64;
        let filled = filled.clamp(0, BAR_LENGTH);
        let empty = BAR_LENGTH - filled;

        let bar: String = "\u{2588}".repeat(filled as usize) + &"\u{2591}".repeat(empty as usize);
        let percentage = score * PERCENTAGE_VALUE;

        let status = if score >= SCORE_THRESHOLD_HIGH {
            "\u{1F7E2} Good"
        } else if score >= SCORE_THRESHOLD_MEDIUM {
            "\u{1F7E1} Fair"
        } else {
            "\u{1F534} Poor"
        };

        format!("{label}: [{bar}] {percentage:.1}% {status}")
    }

    /// Sorts collection rows by the configured key using [`to_comparable`].
    fn sort_collection(&self, rows: &mut [&Item]) {
        let sort_by = &self.config.sort_by;
        let desc = self.config.sort_order == "desc";
        rows.sort_by(|a, b| {
            let ca = to_comparable(a.get(sort_by));
            let cb = to_comparable(b.get(sort_by));
            let ord = ca.partial_cmp(&cb).unwrap_or(std::cmp::Ordering::Equal);
            if desc {
                ord.reverse()
            } else {
                ord
            }
        });
    }
}

/// Extracts every numeric value from a report as `f64`.
///
/// Shared by the formatter summary and [`crate::reporter::Reporter`].
pub(crate) fn extract_all_numeric_metrics(report: &Report) -> BTreeMap<String, f64> {
    let mut metrics = BTreeMap::new();
    for (key, value) in report {
        if let Some(score) = value.to_float64() {
            metrics.insert(key.clone(), score);
        }
    }
    metrics
}

/// Returns every distinct key across a collection's items, sorted (mirrors the
/// Go `getCollectionKeys` + `mapx.SortedKeys`).
fn collection_keys(collection: &[&Item]) -> Vec<String> {
    let mut set = std::collections::BTreeSet::new();
    for item in collection {
        for key in item.keys() {
            set.insert(key.clone());
        }
    }
    set.into_iter().collect()
}

/// Converts a value to a comparable `f64` for sorting, mirroring the Go
/// `toComparable`: numbers map to their value, strings to their byte length, and
/// everything else to `0`.
fn to_comparable(value: Option<&Value>) -> f64 {
    match value {
        Some(Value::Float(n)) => *n,
        Some(Value::Float32(n)) => f64::from(*n),
        Some(Value::Int(n)) => *n as f64,
        Some(Value::Int32(n)) => f64::from(*n),
        Some(Value::String(s)) => s.len() as f64,
        _ => 0.0,
    }
}

/// Renders a table in the `go-pretty` `StyleLight` style with row/column
/// separators, border, and header separators all disabled (matching the
/// `formatter.go` configuration).
///
/// Columns are padded to their widest cell using **byte** length (matching Go's
/// `text.RuneWidth`-free path noted in DESIGN §2.7), cells separated by a single
/// space, with a light box-drawing rule (`─`) under the header and above the
/// footer. This output is NON-BINDING (cosmetic).
pub fn render_style_light_table(header: &[String], body: &[Vec<String>], footer: &str) -> String {
    let cols = header.len();
    let mut widths: Vec<usize> = header.iter().map(|h| h.len()).collect();
    for row in body {
        for (i, cell) in row.iter().enumerate() {
            if i < cols && cell.len() > widths[i] {
                widths[i] = cell.len();
            }
        }
    }

    let pad = |cell: &str, width: usize| -> String {
        let mut s = String::from(cell);
        // Byte-width padding, per DESIGN §2.7.
        if cell.len() < width {
            s.push_str(&" ".repeat(width - cell.len()));
        }
        s
    };

    let render_row = |cells: &[String]| -> String {
        cells
            .iter()
            .enumerate()
            .map(|(i, c)| pad(c, *widths.get(i).unwrap_or(&0)))
            .collect::<Vec<_>>()
            .join(" ")
            .trim_end()
            .to_string()
    };

    let total_width: usize = widths.iter().sum::<usize>() + cols.saturating_sub(1);
    let rule = "\u{2500}".repeat(total_width);

    let mut out: Vec<String> = Vec::with_capacity(body.len() + 3);
    out.push(render_row(header));
    out.push(rule.clone());
    for row in body {
        out.push(render_row(row));
    }
    if !footer.is_empty() {
        out.push(rule);
        out.push(footer.to_string());
    }
    out.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    fn report(pairs: &[(&str, Value)]) -> Report {
        pairs.iter().map(|(k, v)| (k.to_string(), v.clone())).collect()
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_Nil
    #[test]
    fn format_report_nil() {
        let f = Formatter::new(FormatConfig::default());
        assert_eq!(f.format_report(None), "No report data available");
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_Basic
    #[test]
    fn format_report_basic() {
        let f = Formatter::new(FormatConfig::default());
        let r = report(&[
            ("analyzer_name", Value::String("test_analyzer".into())),
            ("message", Value::String("Analysis complete".into())),
        ]);
        let out = f.format_report(Some(&r));
        assert!(out.contains("TEST_ANALYZER"));
        assert!(out.contains("Analysis complete"));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_WithMetrics
    #[test]
    fn format_report_with_metrics() {
        let f = Formatter::new(FormatConfig::default());
        let r = report(&[
            ("analyzer_name", Value::String("test".into())),
            ("score", Value::Float(0.85)),
            ("quality", Value::Float(0.92)),
        ]);
        let out = f.format_report(Some(&r));
        assert!(out.contains("score"));
        assert!(out.contains("quality"));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_SkipHeader
    #[test]
    fn format_report_skip_header() {
        let f = Formatter::new(FormatConfig {
            skip_header: true,
            ..Default::default()
        });
        let r = report(&[
            ("analyzer_name", Value::String("test_analyzer".into())),
            ("message", Value::String("Test".into())),
        ]);
        let out = f.format_report(Some(&r));
        assert!(!out.contains("==="));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_WithProgressBars
    #[test]
    fn format_report_with_progress_bars() {
        let f = Formatter::new(FormatConfig {
            show_progress_bars: true,
            ..Default::default()
        });
        let r = report(&[
            ("analyzer_name", Value::String("test".into())),
            ("score", Value::Float(0.75)),
        ]);
        let out = f.format_report(Some(&r));
        assert!(out.contains("Progress:"));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_WithTables
    #[test]
    fn format_report_with_tables() {
        let f = Formatter::new(FormatConfig {
            show_tables: true,
            ..Default::default()
        });
        let items = vec![
            report(&[("name", Value::String("item1".into())), ("value", Value::Int(10))]),
            report(&[("name", Value::String("item2".into())), ("value", Value::Int(20))]),
        ];
        let r = report(&[
            ("analyzer_name", Value::String("test".into())),
            ("items", Value::Collection(items)),
        ]);
        let out = f.format_report(Some(&r));
        assert!(out.contains("items"));
        assert!(out.contains("Total: 2 items"));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_WithDetails
    #[test]
    fn format_report_with_details() {
        let f = Formatter::new(FormatConfig {
            show_details: true,
            ..Default::default()
        });
        let r = report(&[
            ("analyzer_name", Value::String("test".into())),
            ("detail_field", Value::String("detail_value".into())),
        ]);
        let out = f.format_report(Some(&r));
        assert!(out.contains("Details:"));
    }

    // Ported from formatter_test.go: TestFormatter_CreateProgressBar
    #[test]
    fn create_progress_bar_status() {
        let f = Formatter::new(FormatConfig::default());
        assert!(f.create_progress_bar("test", 0.9).contains("\u{1F7E2} Good"));
        assert!(f.create_progress_bar("test", 0.7).contains("\u{1F7E1} Fair"));
        assert!(f.create_progress_bar("test", 0.3).contains("\u{1F534} Poor"));
    }

    // Ported from formatter_test.go: TestFormatter_FormatReport_SortedDetails
    #[test]
    fn format_report_sorted_details() {
        let f = Formatter::new(FormatConfig {
            show_details: true,
            ..Default::default()
        });
        let r = report(&[
            ("zebra", Value::String("last".into())),
            ("alpha", Value::String("first".into())),
        ]);
        let out = f.format_report(Some(&r));
        let alpha = out.find("alpha").unwrap();
        let zebra = out.find("zebra").unwrap();
        assert!(alpha < zebra);
    }

    // Ported from formatter_test.go: TestFormatter_SortCollection
    #[test]
    fn sort_collection_desc() {
        let f = Formatter::new(FormatConfig {
            show_tables: true,
            sort_by: "value".into(),
            sort_order: "desc".into(),
            ..Default::default()
        });
        let items = vec![
            report(&[("name", Value::String("low".into())), ("value", Value::Int(10))]),
            report(&[("name", Value::String("high".into())), ("value", Value::Int(30))]),
            report(&[("name", Value::String("mid".into())), ("value", Value::Int(20))]),
        ];
        let r = report(&[("items", Value::Collection(items))]);
        let out = f.format_report(Some(&r));
        let high = out.find("high").unwrap();
        let low = out.find("low").unwrap();
        assert!(high < low);
    }
}
